Enable building of msix package

SteveL-MSFT opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

Add support in packaging.psm1 to produce a .msix AppX package.  Update the docker image to use the new msix package type.  Update the associated yml files so AzDevOps performs the build.

## PR Context

Enable publishing PSCore6 to Microsoft Store

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission) Doc PR will be submitted once we build a signed msix
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)

SteveL-MSFT committed:
enable building of msix package

SteveL-MSFT committed:
update packaging to build for msix

SteveL-MSFT committed:
add building of the msix to docker image

SteveL-MSFT committed:
update azdevops ymls to sign and publish msix

adityapatwardhan commented on tools/packaging/packaging.psm1:
```diff
@@ -335,6 +335,24 @@ function Start-PSPackage {
                 }
             }
+            "msix" {
+                $TargetArchitecture = "x64"
```
Do we need `arm64` here?

SteveL-MSFT replied:
As far as I know, the makeappx.exe tool is just creating a pkg so it doesn't matter what architecture the binaries are built for so there's no reason to even check `$TargetArchitecture`.  I'll update it so it will work essentially as a `zip` without needing to know the architecture other than naming the resulting msix.

adityapatwardhan commented on tools/packaging/packaging.psm1:
```diff
@@ -2726,6 +2744,87 @@ function New-MSIPackage
+
+        # Architecture to use when creating the MSI
+        [Parameter(Mandatory = $true)]
+        [ValidateSet("x86", "x64")]
```
arm64?

SteveL-MSFT replied:
Will remove this, since it doesn't really care what the architecture is.

adityapatwardhan commented on tools/packaging/packaging.psm1:
```diff
@@ -2726,6 +2744,87 @@ function New-MSIPackage
+
+    $makeappx = Get-Command makeappx -CommandType Application -ErrorAction Ignore
+    if ($null -eq $makeappx) {
+        $makeappx = Get-ChildItem "c:\makeappx" -Include makeappx.exe -Recurse | Select-Object -First 1
```
hardcoded `c:\`

SteveL-MSFT replied:
@TravisEz13 says that this is only used in our dockerimage which has root path as `c:\`.  Although I guess it wouldn't hurt to change it to `$env:SystemDrive`.

SteveL-MSFT committed:
address Aditya's feedback

SteveL-MSFT commented on tools/releaseBuild/Images/microsoft_powershell_windowsservercore/Dockerfile:
```diff
@@ -15,6 +15,7 @@ RUN Import-Module ./containerFiles/dockerInstall.psm1; `
 
 # Install WIX
 ADD https://github.com/wixtoolset/wix3/releases/download/wix311rtm/wix311-binaries.zip /wix.zip
+ADD https://pscoretestdata.blob.core.windows.net/build-files/makeappx/makeappx.tgz?sp=r&st=2019-04-03T20:37:29Z&se=2020-04-04T04:37:29Z&spr=https&sv=2018-03-28&sig=wGY3gCjVcyowIfGVKWSdiOlkXMrG243YBp7LWrMB2BY%3D&sr=b /makeappx
```
Does renaming to .tar.gzip make it just work or does `ADD` not do what the documentation says?

SteveL-MSFT replied:
After discussion, changing to .zip since it appears that auto-expanding of tgz doesn't work in dockerfile

iSazonov commented on tools/packaging/packaging.psm1:
```diff
@@ -335,6 +335,18 @@ function Start-PSPackage {
+                    Force = $Force
+                }
+
+                if ($PSCmdlet.ShouldProcess("Create MSI Package")) {
```
MSI?

SteveL-MSFT replied:
Will fix

iSazonov commented on assets/AppxManifest.xml:
```diff
@@ -0,0 +1,44 @@
+  </Properties>
+
+  <Dependencies>
+    <TargetDeviceFamily Name="Windows.Universal" MinVersion="10.0.17134.0" MaxVersionTested="10.0.17134.0" />
```
Can we use variables for the versions?

MaxVersionTested  - I'd expect that we tested the package on latest public Windows builds too.

SteveL-MSFT replied:
Let me test it on 1809, not sure when FullTrustApplication was introduced as previously appx only supported running in a sandbox which wouldn't work for pwsh

SteveL-MSFT replied:
Tried it on 1803 which doesn't work (doesn't understand .msix extension).  Works on 1809 which is public.  Updated the xml to reflect 1809.

iSazonov commented on tools/packaging/packaging.psm1:
```diff
@@ -2726,6 +2738,81 @@ function New-MSIPackage
+        [Switch] $Force
+    )
+
+    $makeappx = Get-Command makeappx -CommandType Application -ErrorAction Ignore
```
Make sense check version of the tool?

SteveL-MSFT replied:
I don't expect people to run this locally as the msix must be signed otherwise it won't install.  We're using a specific version of the tool we know works in our docker image to build this.

SteveL-MSFT committed:
address Ilya and Travis' feedback

SteveL-MSFT committed:
specify that the msix only works on 1809, fix searching of makeappx for docker

SteveL-MSFT committed:
makeappx is very specific about version strings, need to add revision
also fixed the search for makeappx when running locally

SteveL-MSFT committed:
enable build for arm/arm64 and update expected artifacts

wjk commented:
We would probably want to add the following XML elements to the manifest:

```xml
<Properties>
  <desktop6:RegistryWriteVirtualization>disabled</desktop6:RegistryWriteVirtualization>
</Properties>
<!-- other content... -->
<Capabilities>
  <rescap:Capability Name="unvirtualizedResources" />
</Capabilities>
```

This will instruct Windows to allow the packaged copy of PowerShell to modify the system Registry directly. Without it, PowerShell's Registry manipulation features will not function properly (the changes will be visible only to PowerShell), vastly impairing the usefulness of the program.

Note that users will need to be running 19H1 for this to work. The use of the desktop6 namespace should not impair loading of the package on downlevel versions of Windows 10 (assuming `IgnorableNamespaces` is used), but the capability might.

SteveL-MSFT commented on tools/releaseBuild/azureDevOps/templates/upload.yml:
```diff
@@ -34,3 +35,19 @@ steps:
+- template: upload-final-results.yml
+  parameters:
+    artifactPath: $(Build.StagingDirectory)\signedPackages
+    artifactFilter: PowerShell-${{ parameters.version }}-win-${{ parameters.architecture }}.msix
```
Yes, will have to update this.

SteveL-MSFT replied:
I'm just going to have the package name match what we call the msi package.  It's only the package version that will differ due to requirements for appx.

SteveL-MSFT commented:
@wjk Thanks for that snippet!  That should resolve the issue I see with using the registry provider from the msix.  Will update the manifest.

SteveL-MSFT commented:
@wjk With those additions, it doesn't fix some issues with the msix package in that `Move-Item`, `Copy-Item` and `Rename-Item` doesn't work with registry provider when elevated and against HKLM:.  `New-Item` and `Remove-Item` do work.  And all cmdlets seem to work against HKCU:.

SteveL-MSFT committed:
add capability to not use virtualized resources
fix packaging to create resources.pri so that taskbar icon doesn't have background border
fix naming of package to align with msi package

SteveL-MSFT committed:
address scriptanalyzer issue

wjk commented:
@SteveL-MSFT Unfortunately, I'm not running 19H1 yet, so I can’t help debug your HKLM issue. It seems really weird that some of the registry cmdlets would work, but others don’t. Do you see an error message, or do the changes simply silently fail?

SteveL-MSFT commented:
@wjk Getting an `Access Denied` error when attempting a move, rename, or copy.  Here's the stacktace:

```powershell
PSMessageDetails      :
Exception             : System.Security.SecurityException: Requested registry access is not allowed.
                           at Microsoft.Win32.ThrowHelper.ThrowSecurityException(String msg)
                           at Microsoft.Win32.RegistryKey.InternalOpenSubKeyCore(String name, Boolean writable,
                        Boolean throwOnPermissionFailure)
                           at Microsoft.PowerShell.Commands.RegistryWrapper.OpenSubKey(String name, Boolean writable)
                           at Microsoft.PowerShell.Commands.RegistryProvider.GetRegkeyForPath(String path, Boolean
                        writeAccess)
                           at Microsoft.PowerShell.Commands.RegistryProvider.CopyRegistryKey(IRegistryWrapper key,
                        String path, String destination, Boolean recurse, Boolean streamResult, Boolean
                        streamFirstOnly)
                           at Microsoft.PowerShell.Commands.RegistryProvider.CopyItem(String path, String destination,
                        Boolean recurse)
TargetObject          : HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\PowerShell\3\test\test
CategoryInfo          : PermissionDenied: (HKEY_LOCAL_MACHINE\\u2026erShell\3\test\test:String) [Copy-Item],
                        SecurityException
FullyQualifiedErrorId : System.Security.SecurityException,Microsoft.PowerShell.Commands.CopyItemCommand
ErrorDetails          :
InvocationInfo        : System.Management.Automation.InvocationInfo
ScriptStackTrace      : at <ScriptBlock>, <No file>: line 1
PipelineIterationInfo : {0, 1}
```

SteveL-MSFT committed:
update blog to include makepri

SteveL-MSFT committed:
enable build for arm32 and arm64

SteveL-MSFT committed:
add arm32 and arm64 to signing

SteveL-MSFT committed:
spell Authenticode correctly

SteveL-MSFT commented on assets/AppxManifest.xml:
```diff
@@ -0,0 +1,49 @@
+  </Properties>
+
+  <Dependencies>
+    <TargetDeviceFamily Name="Windows.Universal" MinVersion="10.0.17763.0" MaxVersionTested="10.0.17763.0" />
```
I tried the msix on 1803 and it didn't install understand msix, I think it only supports .appx.

SteveL-MSFT replied:
We can update the version information once we verify it'll work on those versions.

adityapatwardhan commented on tools/releaseBuild/Images/microsoft_powershell_windowsservercore/PowerShellPackage.ps1:
```diff
@@ -122,6 +122,13 @@ try{
         Start-PSPackage @pspackageParams @releaseTagParam
     }
 
+    if (!$ComponentRegistration.IsPresent -and !$Symbols.IsPresent -and $Runtime -notin "win7-x86",'fxdependent')
```
nit: use either double quotes or single quotes, single quotes preffered.

SteveL-MSFT replied:
will update

SteveL-MSFT committed:
address Aditya's feedback

PoshChan commented:
@SteveL-MSFT, your last commit had 1 failures in `PowerShell-CI-macos`
Validate 'Save-Help -DestinationPath for all PowerShell Core modules..Validate Save-Help for the 'Microsoft.PowerShell.Core' module
```powershell
Expected $null, but got 'Microsoft.PowerShell.Core_00000000-0000-0000-0000-000000000000_en-US_HelpContent.zip'.
at <ScriptBlock>, /Users/vsts/agent/2.149.2/work/1/s/test/powershell/engine/Help/UpdatableHelpSystem.Tests.ps1: line 332
332:     $expectedCompressedFile | Should -Be $compressedFile
```

SteveL-MSFT commented:
@PoshChan please retry macos

PoshChan commented:
@SteveL-MSFT, successfully started retry of `PowerShell-CI-macOS`
